include/deal.II/multigrid/mg_constrained_dofs.h

```diff
@@ -146,13 +169,64 @@ MGConstrainedDoFs::initialize(const DoFHandler<dim,spacedim> &dof)
 {
   boundary_indices.clear();
   refinement_edge_indices.clear();
+  level_constraints.clear();
 
   const unsigned int nlevels = dof.get_triangulation().n_global_levels();
 
-  //At this point refinement_edge_indices is empty.
+  //At this point level_constraint and refinement_edge_indices are empty.
+  level_constraints.resize(nlevels);
   refinement_edge_indices.resize(nlevels);
   for (unsigned int l=0; l<nlevels; ++l)
-    refinement_edge_indices[l] = IndexSet(dof.n_dofs(l));
+    {
+      IndexSet relevant_dofs;
+      DoFTools::extract_locally_relevant_level_dofs(dof,l,relevant_dofs);
+      level_constraints[l].reinit(relevant_dofs);
+
+      // Loop through relevant cells and faces finding those which are periodic neighbors.
+      typename DoFHandler<dim,spacedim>::cell_iterator
+      cell = dof.begin(l),
+      endc = dof.end(l);
+      for (; cell!=endc; ++cell)
+        {
+          if (cell->level_subdomain_id() == numbers::artificial_subdomain_id)
+            continue;
+
+          for (unsigned int f = 0;
+               f < GeometryInfo<dim>::faces_per_cell;
+               ++f)
+            if (cell->has_periodic_neighbor(f))
+              {
+                if (cell->level_subdomain_id() ==  dof.get_triangulation().locally_owned_subdomain())
+                  {
+                    Assert(cell->periodic_neighbor(f)->level_subdomain_id() != numbers::artificial_subdomain_id,
+                           ExcMessage ("Periodic neighbor of a locally owned cell must either be owned or ghost."));
+                  }
+                else if (cell->periodic_neighbor(f)->level_subdomain_id() == numbers::artificial_subdomain_id)
+                  {
+                    continue;
```

Comment:
can you add a comment here to say that "cell is a level-ghost and its neighbor is a level-artificial cell -- nothing to do here"

I think it might be worth checking the first part of this statement with `Assert (cell->is_locally_owned_on_level() == false, ExcInternalError());`